### PR TITLE
Clean up classifier code, add better metrics

### PR DESCRIPTION
**Pull request by kbrose**
* [ ] improve performance / stability
* [x] better metrics tracking (accuracy per class, put in tensorboard for better visualization)
* [x] model save / load functionality
* [x] standardize selection of classes -- I manually moved the classes with < 1000 images out of the folder (except the "empty" class), but this should probably be a parameter

### CODE REVIEW

**Comment by kbrose** [comment]
The true positive rate and false positive rate of each class is now printed out during training:

```
100%|####################################################################| 1617/1617 [00:48<00:00, 33.29it/s]
Epoch    8, Loss: 0.28, Accuracy:  91.22%, Test Loss: 1.43, Test Accuracy:  64.12%
             TPR     FPR   (prevelance)
      bird  73.02%   5.38% (     6.23%)
    bobcat  54.80%   4.86% (     7.81%)
       car  99.62%   0.91% (     3.95%)
       cat  33.53%   4.54% (     6.79%)
    coyote  78.04%   9.55% (    10.06%)
      deer  93.18%   8.80% (     9.94%)
       dog  49.35%   3.25% (     4.42%)
       fox  20.96%   1.89% (     1.70%)
   opossum  45.71%  10.55% (    18.76%)
    rabbit  66.91%   3.45% (     9.32%)
   raccoon  80.32%  10.19% (    12.23%)
    rodent  21.49%   2.05% (     2.85%)
     skunk  52.92%   1.43% (     1.85%)
  squirrel  79.42%   3.93% (     4.09%)
```

I've also added tensorboard metrics (not sure how that would work with colab). You can run the tensorboard server with `tensorboard --logdir=.tensorboard`

image

**Comment by lukeforehand** [comment]
@kbrose I have some commits that get tensorboard working in colab, they made it pretty easy to integrate :-)

https://github.com/lukeforehand/deertracker/blob/kbrose-classifier-cleanup/training.ipynb

What are your recommendation for improving accuracy?  I am also wondering how we should go about combining training datasets, like the whitetail dataset + caltech.  I used a superset of the caltech classes in the whitetail dataset labeling.

**Comment by lukeforehand** [comment]
I'm happy to go on gathering photos and labeling!  I tweaked tkteach to be even faster and I was going about one image per 1.2 seconds.

**Comment by lukeforehand** [comment]
tensorboard

**Comment by kbrose** [comment]
> What are your recommendation for improving accuracy? ... I'm happy to go on gathering photos and labeling!

More data is usually helpful certainly.

Something else you could play with is different base models. The code currently uses MobileNet which was designed for speed (i.e. to work well on phones) at a slight cost to accuracy. You could try replacing that with some of the other options in [`tf.keras.applications`](https://www.tensorflow.org/api_docs/python/tf/keras/applications) like one of the DenseNets or ResNets. The larger versions of those models perform better at the cost of peak memory usage (both your main RAM and the GPU's on-board RAM) and speed.

A caveat is that while the bigger models do better on the task they were trained on, it doesn't necessarily translate to better performance on the new task.

Another thing that can sometimes improve performance is reduce the overall number of classes so that the model can focus better.

> I am also wondering how we should go about combining training datasets, like the whitetail dataset + caltech. I used a superset of the caltech classes in the whitetail dataset 

I think it'd be fine to combine the datasets directly, except for the cases where you replaced the class entirely. I.e. don't keep caltech's "deer" class if you're going to add specific doe/buck/etc. classes since that will make the model fight to decide between them.

**Comment by kbrose** [comment]
Latest commit will save the model every epoch that has the best testing loss. The save "format" is a full directory containing several subfolders/files:

```bash
$ tree v0-0004  # format is {model_name}-{epoch_number}
v0-0004
├── assets
├── saved_model.pb
└── variables
    ├── variables.data-00000-of-00001
    └── variables.index
```

Loading and predicting is then pretty straightforward:

```python
import tensorflow as tf
import numpy as np

model = tf.keras.models.load_model('models/v0/v0-0004/')
# models are trained with an extra batch dimension, during inference
# we typically pass in just one example at a time, so add an extra
# singular dimension to the input, and index it from the output
prediction = model(np.random.rand(1, 96, 96, 3)).numpy()[0]

# alternate way to load if you don't want to remember the tf function path
from deertracker.classifier import load_model
model = load_model(path)
```

**Comment by kbrose** [comment]
@lukeforehand I think this is probably good to merge. I haven't done anything for the performance/stability bullet, but that's 95% of the art of doing ML so it probably won't get done in one MR.

**Comment by lukeforehand** [comment]
> Another thing that can sometimes improve performance is reduce the overall number of classes so that the model can focus better.

When we prune out classes with a small number of images in the beginning, is that reducing the overall classes?

> I think it'd be fine to combine the datasets directly, except for the cases where you replaced the class entirely. I.e. don't keep caltech's "deer" class if you're going to add specific doe/buck/etc. classes since that will make the model fight to decide between them.

Ok I might label the caltech deer images more specifically then, the labeling tool is very quick.

**Comment by lukeforehand** [comment]
> I think it'd be fine to combine the datasets directly, except for the cases where you replaced the class entirely. I.e. don't keep caltech's "deer" class if you're going to add specific doe/buck/etc. classes since that will make the model fight to decide between them.

Also I notice in training with the "whitetail" dataset, the buck and doe classes might be fighting.  I think there is too much similarity in a lot of the training data, some of the buck photos the antlers are barely noticable.

**Comment by kbrose** [comment]
> When we prune out classes with a small number of images in the beginning, is that reducing the overall classes?

Yep!

> Also I notice in training with the "whitetail" dataset, the buck and doe classes might be fighting. I think there is too much similarity in a lot of the training data, some of the buck photos the antlers are barely noticable, but the body and neck are clearly different.

Yeah, I'd believe it struggles with that. I'll put some thought into how we might make that better, but for now the best way probably is just more examples.

> I've become an expert after labeling 3000 images :-D

That's secretly one of my favorite parts of doing ML work. Becoming an expert at an extremely specific and mostly useless repetitive tasks. Makes for good audio book / podcast listening time :)

**Comment by lukeforehand** [comment]
This is some amazing work, I'm very much appreciating your help in this!
